### ISSUE TITLE
Unable to connect to a Fabric server. / How to register a recipe serialiser and command node parser?

### ISSUE THREAD

**Issue by ajh123** [adepierre/Botcraft#124]
I'm trying to connect a bot to my Fabric server which is behind a Velcoity proxy.

When the bot joins the proxy it uses the `/server ...` command to switch to the Fabric server. The bot is able to connect but it instantly disconnects.

In the Fabric server's logs I see:
```java
[13:58:36] [Server thread/INFO]: bob[/10.0.0.160:45864] logged in with entity id 212 at (0.5, 1.0, 8.5)
[13:58:36] [Server thread/INFO]: bob joined the game
[13:58:36] [Server thread/INFO]: Updating all forceload tickets for 8e289159-2034-3a16-96b9-9fa637848b3b
[13:58:36] [Server thread/INFO]: bob lost connection: Disconnected
[13:58:36] [Server thread/INFO]: bob left the game
```

In the Velocity server's logs I see:

```java
[13:58:31] [Velocity Async Event Executor - #1/INFO] [com.velocitypowered.proxy.VelocityServer]: UUID of player bob is 8e289159-2034-3a16-96b9-9fa637848b3b
[13:58:31] [Netty epoll Worker #2/INFO] [com.velocitypowered.proxy.connection.MinecraftConnection]: [server connection] bob -> Lobby_1 has connected
[13:58:36] [Netty epoll Worker #2/INFO] [com.velocitypowered.proxy.connection.MinecraftConnection]: [server connection] bob -> ForgeCreative has connected // this is actually the Fabric server
[13:58:36] [Netty epoll Worker #2/INFO] [com.velocitypowered.proxy.connection.MinecraftConnection]: [server connection] bob -> Lobby_1 has disconnected
[13:58:36] [Netty epoll Worker #2/INFO] [com.velocitypowered.proxy.connection.client.ConnectedPlayer]: [connected player] bob (/10.0.0.160:56333) has disconnected: An internal error occurred in your connection.
[13:58:36] [Netty epoll Worker #2/ERROR] [com.velocitypowered.proxy.connection.MinecraftConnection]: [connected player] bob (/10.0.0.160:56333): exception encountered in com.velocitypowered.proxy.connection.client.ClientPlaySessionHandler@8a72772
io.netty.channel.unix.Errors$NativeIoException: recvAddress(..) failed: Connection reset by peer
[13:58:36] [Netty epoll Worker #2/INFO] [com.velocitypowered.proxy.connection.MinecraftConnection]: [server connection] bob -> ForgeCreative has disconnected  // this is actually the Fabric server
```

The logs of the bot says:
```java
$ ../bin/miners_online_bot_code_d --address 10.0.0.160:3000 --login bob
[2024-01-16 13:58:30.384] [INFO] [main(140646151354432)] main.cpp(63): Starting connection process
[2024-01-16 13:58:30.384] [INFO] [main(140646151354432)] TCP_Com.cpp(31): Trying to connect to 10.0.0.160:3000
[2024-01-16 13:58:30.385] [INFO] [NetworkIOService(140646142944832)] TCP_Com.cpp(96): Connection to server established.
[2024-01-16 13:58:36.284] [FATAL] [NetworkPacketProcessing - bob(140646151337536)] NetworkManager.cpp(338): Exception: Not enough input in ReadData<std::string>
terminate called after throwing an instance of 'std::runtime_error'
  what():  Not enough input in ReadData<std::string>
Aborted
```

I've debugged further using Sniffcraft and it says:
```java
$ ./sniffcraft-windows-1.20.1.exe 3000 10.0.0.97:25575
Ready to redirect any client connection on 127.0.0.1:3000 to 10.0.0.97:25575
Loading updated conf file...
Conf file loaded!
Starting new proxy to 10.0.0.97:25575
Ready to redirect any client connection on 127.0.0.1:3000 to 10.0.0.97:25575
Server --> Client: PARSING EXCEPTION for message Update Recipes(: 109)Not enough input in ReadData<std::string>
Server --> Client: PARSING EXCEPTION for message Commands(: 16)Unable to create command node parser with id: 51. Are you using a mod? Modded protocol is not supported.
```

When I join the Fabric server with a Fabric client that does not have other mods installed, Minecraft has the error:

![image](https://github.com/adepierre/Botcraft/assets/41990982/cbdfb966-a891-49ac-8953-76c265c59b75)

This seems like its the cause for the "Server --> Client: PARSING EXCEPTION for message Update Recipes(: 109)Not enough input in ReadData<std::string>" in Sniffcraft. 

**TLDR:**

_If so_, how can I implement my own recipe serialiser? Also, how can I register my own command node parser? I'm happy to do the code myself, I just need to know what to use.

**Comment by adepierre** [comment]
Hello, and thank you for the detailed report.

The error you have seems indeed to be caused by non vanilla recipes and command node.

To add custom recipes parser, you need to create a class in [protocolCraft/include/protocolCraft/Types/Recipes](https://github.com/adepierre/Botcraft/tree/master/protocolCraft/include/protocolCraft/Types/Recipes) (you can check RecipeTypeDataStoneCutting for reference). Then you need to register it with the correct identifier in the if of the function [CreateRecipeTypeData](https://github.com/adepierre/Botcraft/blob/master/protocolCraft/src/Types/Recipes/RecipeTypeData.cpp#L25). You can have multiple identifiers matching the same class if all recipes type have the same underlying data in the same order.

For command nodes, the process is similar, with a little difference as they are identified by ID instead of strings. Data layout is defined by a class in [protocolCraft/include/protocolCraft/Types/CommandNode](https://github.com/adepierre/Botcraft/tree/master/protocolCraft/include/protocolCraft/Types/CommandNode), and the matching ID needs to be added to the [BrigadierPropertyType enum](https://github.com/adepierre/Botcraft/blob/master/protocolCraft/include/protocolCraft/Types/CommandNode/BrigadierProperty.hpp#L13). Factory to link ID and data class is [here](https://github.com/adepierre/Botcraft/blob/master/protocolCraft/src/Types/CommandNode/BrigadierProperty.cpp#L83). Once added to the switch case, protocolCraft should be able to handle the new node types.

**Comment by ajh123** [comment]
Thanks! That should be all I need to do.
